File: javascript/src/recording.js
import {BinaryReader} from "google-protobuf";
import assert from "assert";

function load_recording(data) {
    let recording = {
        begin_time: null,
        end_time: null,
        duration: null,
        channels: [],
    };

    let reader = new BinaryReader(data);
    let first = true;
    while(reader.getCursor() < data.length) {
        let event = new proto.indoors.pipeline.protocol.Event();
        reader.nextWireType_ = 2; // TODO hack and BinaryConstants.WireType.DELIMITED;
        reader.readMessage(event, proto.indoors.pipeline.protocol.Event.deserializeBinaryFromReader);

        if (first) {
            assert(event.hasHello(), 'expecting hello as first message');

            recording.begin_time = event.getT();

            event.getHello().getChannelsList().forEach((channel) => {
                let mapping = {
                    annotation: channel.getAnnotation(),
                    t: [],
                };

                if (channel.getEventexample().hasClock())
                    mapping = Object.assign(mapping, {unixUtc: []});
                else if (channel.getEventexample().hasVector2())
                    mapping = Object.assign(mapping, {x: [], y: []});
                else if (channel.getEventexample().hasVector3())
                    mapping = Object.assign(mapping, {x: [], y: [], z: []});
                else if (channel.getEventexample().hasQuaternion())
                    mapping = Object.assign(mapping, {w: [], x: [], y: [], z: []});

                recording.channels[channel.getChannel()] = mapping;
            });

            first = false;
        } else if (event.hasBye()) {
            recording.end_time = event.getT();
            recording.duration = recording.end_time - recording.begin_time;
            break;
        } else {
            let mapping = recording.channels[event.getChannel()];
            mapping.t.push(event.getT());

            if (event.hasClock()) {
                mapping.unixUtc.push(event.getClock().getUnixUtc());
            } else if (event.hasVector2()) {
                mapping.x.push(event.getVector2().getX());
                mapping.y.push(event.getVector2().getY());
            } else if (event.hasVector3()) {
                mapping.x.push(event.getVector3().getX());
                mapping.y.push(event.getVector3().getY());
                mapping.z.push(event.getVector3().getZ());
            } else if (event.hasQuaternion()) {
                mapping.w.push(event.getQuaternion().getW());
                mapping.x.push(event.getQuaternion().getX());
                mapping.y.push(event.getQuaternion().getY());
                mapping.z.push(event.getQuaternion().getZ());
            }
        }
    }

    return recording;
}

function load_recording_from_url(url, callback) {
    let req = new XMLHttpRequest();
    req.open("GET", url, true);
    req.responseType = "arraybuffer";

    req.onload = () => {
        if (!req.response) {
            callback(null);
            return;
        }

        let data = new Uint8Array(req.response);
        let recording = load_recording(data);

        callback(recording);
    };

    req.send();
}

export { load_recording, load_recording_from_url }

File: python/recording.py
#!/usr/bin/env python

from indoors.pipeline.protocol import event_pb2
from google.protobuf.internal.decoder import _DecodeVarint32
import pandas as pd

def load(path):
    result = {
        'begin_time': None,
        'end_time': None,
        'duration': None,
        'channels': {},
    }

    with open(path, 'rb') as f:
        first = True
        while True:
            len_data = f.read(4)
            if not len_data:
                break
            msg_len, len_len = _DecodeVarint32(len_data, 0)
            f.seek(len_len - 4, 1)

            msg = event_pb2.Event()
            msg.ParseFromString(f.read(msg_len))

            if first:
                assert(msg.HasField('hello'))

                result['begin_time'] = msg.t

                for channel in msg.hello.channels:
                    result['channels'][channel.channel] = {
                        'annotation': channel.annotation,
                        'data': [],
                    }

                first = False
                continue

            assert(not msg.HasField('hello'))
            if msg.HasField('bye'):
                break
            event = {'t': msg.t}
            if msg.HasField('clock'):
                event['unixUtc'] = msg.clock.unix_utc
            elif msg.HasField('vector2'):
                event['x'] = msg.vector2.x
                event['y'] = msg.vector2.y
            elif msg.HasField('vector3'):
                event['x'] = msg.vector3.x
                event['y'] = msg.vector3.y
                event['z'] = msg.vector3.z
            elif msg.HasField('quaternion'):
                event['w'] = msg.quaternion.w
                event['x'] = msg.quaternion.x
                event['y'] = msg.quaternion.y
                event['z'] = msg.quaternion.z
            result['channels'][msg.channel]['data'].append(event)

    for k, v in result['channels'].items():
        v['data'] = pd.DataFrame(v['data'])

    return result
